skopt/learning/gaussian_process/gpr.py

```diff
@@ -197,3 +204,112 @@ def fit(self, X, y):
+            The gradient of the predicted mean
+        y_std_grad: shape = (n_samples, n_features)
+            The gradient of the predicted std.
+        """
+        if return_std and return_cov:
+            raise RuntimeError(
+                "Not returning standard deviation of predictions when "
+                "returning full covariance.")
+        if return_std_grad and not return_std:
+            raise ValueError(
+                "Not returning std_gradient without returning "
+                "the mean.")
+        X = check_array(X)
+        if X.shape[0] != 1 and (return_mean_grad or return_std_grad):
+            raise ValueError("Not implemented for n_samples > 1")
+
+        if not hasattr(self, "X_train_"):  # Unfitted;predict based on GP prior
+            y_mean = np.zeros(X.shape[0])
+            if return_cov:
+                y_cov = self.kernel(X)
+                return y_mean, y_cov
+            elif return_std:
+                y_var = self.kernel.diag(X)
+                return y_mean, np.sqrt(y_var)
+            else:
+                return y_mean
+        else:  # Predict based on GP posterior
+            K_trans = self.kernel_(X, self.X_train_)
+            y_mean = K_trans.dot(self.alpha_)  # Line 4 (y_mean = f_star)
+            y_mean = self.y_train_mean + y_mean  # undo normal.
+
+            if return_cov:
+                v = cho_solve((self.L_, True), K_trans.T)  # Line 5
+                y_cov = self.kernel_(X) - K_trans.dot(v)  # Line 6
+                return y_mean, y_cov
+
+            elif return_std:
+                # compute inverse K_inv of K based on its Cholesky
+                # decomposition L and its inverse L_inv
+                L_inv = solve_triangular(self.L_.T, np.eye(self.L_.shape[0]))
+                K_inv = L_inv.dot(L_inv.T)
+                # Compute variance of predictive distribution
+                y_var = self.kernel_.diag(X)
+                y_var -= np.einsum("ki,kj,ij->k", K_trans, K_trans, K_inv)
+
+                # Check if any of the variances is negative because of
+                # numerical issues. If yes: set the variance to 0.
+                y_var_negative = y_var < 0
+                if np.any(y_var_negative):
+                    warnings.warn("Predicted variances smaller than 0. "
+                                  "Setting those variances to 0.")
+                    y_var[y_var_negative] = 0.0
+                y_std = np.sqrt(y_var)
+
+            if return_mean_grad:
+                grad = self.kernel_.gradient_X(X[0], self.X_train_)
+                grad_wrt_mean = np.dot(grad.T   , self.alpha_)
+
+                if return_std_grad:
+                    # XXX: Cache np.dot(K_trans, K_inv) from above
+                    grad_wrt_std = -np.dot(K_trans, np.dot(K_inv, grad))[0] / y_std
+                    return y_mean, y_std, grad_wrt_mean, grad_wrt_std
+                return y_mean, y_std, grad_wrt_mean
```

Comment:
Why do you return `y_std`? This avalanche of flags in the return values (which are not necessarily compatible as far I understand) makes the code very difficult to follow :/